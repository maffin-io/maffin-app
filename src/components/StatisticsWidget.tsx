import React from 'react';

export type StatisticsWidgetProps = {
  title: string,
  stats: string,
  description: string,
  className?: string,
  statsTextClass?: string,
};

export default function StatisticsWidget({
  title,
  stats,
  description,
  className = '',
  statsTextClass = '',
}: StatisticsWidgetProps): JSX.Element {
  return (
    <div className={`${className} bg-gunmetal-700 rounded-sm p-6`}>
      <p>
        {title}
      </p>
      <p className={`${statsTextClass} text-2xl font-semibold my-3`}>{stats}</p>

      <span className="text-sm">
        {description}
      </span>
    </div>
  );
}
