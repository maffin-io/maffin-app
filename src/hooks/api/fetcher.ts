export default function fetcher(f: () => Promise<any>, key: unknown) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${JSON.stringify(key)}: ${end - start}ms`);
    return r;
  };
}
