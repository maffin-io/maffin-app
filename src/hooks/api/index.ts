import { DateTime } from 'luxon';
import { SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity, Price } from '@/book/entities';
import { PriceDB, PriceDBMap } from '@/book/prices';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import getUser from '@/lib/getUser';
import useGapiClient from '@/hooks/useGapiClient';
import type { AccountsMap } from '@/types/book';
import type { Split, Transaction } from '@/book/entities';
import type { User } from '@/types/user';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';
  return useSWRImmutable(
    key,
    // Needs to be encapsulated in arrow function or it doesnt work properly
    fetcher(async () => Commodity.find(), key),
  );
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';
  return useSWRImmutable(
    key,
    fetcher(queries.getAccounts, key),
  );
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: todayPrices } = useTodayPrices();

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    key,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, todayPrices as PriceDBMap),
      key,
    ),
  );

  return result;
}

export function useInvestments(guid?: string): SWRResponse<InvestmentAccount[]> {
  let key = '/api/investments';
  if (guid) {
    key = `${key}/${guid}`;
  }
  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(guid), key),
  );
  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(guid: string): SWRResponse<Price[]> {
  const key = `/api/prices/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(
      async () => Price.find({
        where: {
          fk_commodity: {
            guid,
          },
        },
        order: {
          date: 'ASC',
        },
      }),
      key,
    ),
  );
}

export function useTodayPrices(): SWRResponse<PriceDBMap> {
  const key = '/api/prices/today';
  return useSWRImmutable(
    key,
    fetcher(PriceDB.getTodayQuotes, key),
  );
}

export function useUser(): SWRResponse<User> {
  const [isGapiLoaded] = useGapiClient();
  const key = '/api/user';
  return useSWRImmutable(
    isGapiLoaded ? key : null,
    fetcher(getUser, key),
    {
      refreshInterval: 100000,
      revalidateOnMount: true,
    },
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
