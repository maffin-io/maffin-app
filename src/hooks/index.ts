export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
