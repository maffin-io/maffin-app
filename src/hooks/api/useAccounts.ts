import {
  useQuery,
  UseQueryResult,
} from '@tanstack/react-query';

import { Account } from '@/book/entities';
import fetcher from './fetcher';

export function useAccount(guid: string): UseQueryResult<Account | undefined> {
  const result = useAccounts<Account | undefined>(
    (data => data.find(a => a.guid === guid)),
  );

  return result;
}

export function useAccounts<TData = Account[]>(
  select?: (data: Account[]) => TData,
): UseQueryResult<TData> {
  const result = useQuery({
    queryKey: [...Account.CACHE_KEY],
    queryFn: fetcher(
      async () => {
        const accounts = await Account.find();
        const root = accounts.find(a => a.type === 'ROOT') as Account;
        setAccountPaths(root, accounts);
        return accounts;
      },
      Account.CACHE_KEY,
    ),
    select,
    networkMode: 'always',
  });

  return result;
}

function setAccountPaths(current: Account, accounts: Account[]) {
  const parent = accounts.find(a => a.guid === current.parentId);
  if (!parent || parent.type === 'ROOT') {
    current.path = current.name;
  } else {
    current.path = `${parent.path}:${current.name}`;
  }

  current.childrenIds.forEach(childId => {
    const account = accounts.find(a => a.guid === childId) as Account;
    setAccountPaths(account, accounts);
  });
}
