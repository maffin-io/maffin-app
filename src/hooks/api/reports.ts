import React from 'react';
import {
  useQuery,
} from '@tanstack/react-query';
import { DateTime, Interval } from 'luxon';
import type { UseQueryResult } from '@tanstack/react-query';

import { aggregateChildrenTotals } from '@/helpers/accountsTotalAggregations';
import { Split } from '@/book/entities';
import { getAccountsTotals } from '@/lib/queries';
import { isAsset, isLiability } from '@/book/helpers';
import type { AccountsTotals } from '@/types/book';
import type { Account } from '@/book/entities';
import type { PriceDBMap } from '@/book/prices';
import { useAccounts } from './useAccounts';
import { usePrices } from './usePrices';
import fetcher from './fetcher';

/**
 * Calculates the balance sheet for all asset/liability/equity accounts
 * at a given point in time.
 */
export function useBalanceSheet(
  selectedDate: DateTime = DateTime.now(),
  select?: (data: AccountsTotals) => AccountsTotals,
): UseQueryResult<AccountsTotals> {
  const { data: accounts, dataUpdatedAt: accountsUpdatedAt } = useAccounts();
  const { data: prices, dataUpdatedAt: pricesUpdatedAt } = usePrices({});

  const aggregate = React.useCallback(
    ((data: AccountsTotals) => aggregateChildrenTotals(
      ['type_asset', 'type_liability'],
      accounts as Account[],
      prices as PriceDBMap,
      selectedDate,
      data,
    )),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [accountsUpdatedAt, pricesUpdatedAt, selectedDate.toISODate()],
  );

  const queryKey = [
    ...Split.CACHE_KEY,
    {
      aggregation: 'balance-sheet',
      date: selectedDate.toISODate(),
      accountsUpdatedAt,
    },
  ];
  const result = useQuery({
    queryKey,
    queryFn: fetcher(
      async () => getAccountsTotals(
        (accounts as Account[]).filter(a => a.type !== 'INCOME' && a.type !== 'EXPENSE'),
        Interval.fromDateTimes(
          DateTime.fromISO('1970'),
          selectedDate,
        ),
      ),
      queryKey,
    ),
    enabled: !!accounts,
    select: select || aggregate,
    networkMode: 'always',
  });

  return result;
}

/**
 * Calculates the income statement for all income/expense accounts
 * in a given interval of time.
 */
export function useIEStatement(
  interval: Interval,
  select?: (data: AccountsTotals) => AccountsTotals,
): UseQueryResult<AccountsTotals> {
  const { data: accounts, dataUpdatedAt: accountsUpdatedAt } = useAccounts();
  const { data: prices, dataUpdatedAt: pricesUpdatedAt } = usePrices({});

  const aggregate = React.useCallback(
    ((data: AccountsTotals) => aggregateChildrenTotals(
      ['type_income', 'type_expense'],
      accounts as Account[],
      prices as PriceDBMap,
      interval.end as DateTime<true>,
      data,
    )),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [accountsUpdatedAt, pricesUpdatedAt, interval.toISODate()],
  );

  const queryKey = [
    ...Split.CACHE_KEY,
    {
      aggregation: 'income-statement',
      interval: interval.toISODate(),
      accountsUpdatedAt,
    },
  ];
  const result = useQuery({
    queryKey,
    queryFn: fetcher(
      async () => getAccountsTotals(
        (accounts as Account[]).filter(
          a => !isAsset(a) && !isLiability(a) && a.type !== 'EQUITY',
        ),
        interval,
      ),
      queryKey,
    ),
    enabled: !!accounts,
    select: select || aggregate,
    networkMode: 'always',
  });

  return result;
}

export function useAccountsTotals(
  interval: Interval,
  select?: (data: AccountsTotals) => AccountsTotals,
): UseQueryResult<AccountsTotals> {
  const { data: balanceSheet, dataUpdatedAt: bsUpdatedAt } = useBalanceSheet(
    interval.end as DateTime<true>,
    select,
  );
  const { data: incomeStatement, dataUpdatedAt: iesUpdatedAt } = useIEStatement(interval, select);

  const queryKey = [
    ...Split.CACHE_KEY,
    {
      aggregation: 'total',
      interval: interval.toISODate(),
      bsUpdatedAt,
      iesUpdatedAt,
    },
  ];
  const result = useQuery({
    queryKey,
    queryFn: () => ({ ...balanceSheet, ...incomeStatement }),
    enabled: !!balanceSheet && !!incomeStatement,
    networkMode: 'always',
  });

  return result;
}
