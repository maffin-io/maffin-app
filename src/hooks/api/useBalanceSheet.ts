import React from 'react';
import { DateTime, Interval } from 'luxon';
import { useQuery, UseQueryResult } from '@tanstack/react-query';

import type { AccountsTotals } from '@/types/book';
import { aggregateChildrenTotals } from '@/helpers/accountsTotalAggregations';
import { Split } from '@/book/entities';
import type { Account } from '@/book/entities';
import type { PriceDBMap } from '@/book/prices';
import { getAccountsTotals } from '@/lib/queries';
import { useAccounts } from './useAccounts';
import { usePrices } from './usePrices';
import fetcher from './fetcher';

/**
 * Calculates the balance sheet for all asset/liability/equity accounts
 * at a given point in time.
 */
export function useBalanceSheet(
  selectedDate: DateTime = DateTime.now(),
  select?: (data: AccountsTotals) => AccountsTotals,
): UseQueryResult<AccountsTotals> {
  const { data: accounts, dataUpdatedAt: accountsUpdatedAt } = useAccounts();
  const { data: prices, dataUpdatedAt: pricesUpdatedAt } = usePrices({});

  const aggregate = React.useCallback(
    ((data: AccountsTotals) => aggregateChildrenTotals(
      ['type_asset', 'type_liability'],
      accounts as Account[],
      prices as PriceDBMap,
      selectedDate,
      data,
    )),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [accountsUpdatedAt, pricesUpdatedAt, selectedDate.toISODate()],
  );

  const queryKey = [
    ...Split.CACHE_KEY,
    {
      aggregation: 'balance-sheet',
      date: selectedDate.toISODate(),
      accountsUpdatedAt,
    },
  ];
  const result = useQuery({
    queryKey,
    queryFn: fetcher(
      async () => getAccountsTotals(
        (accounts as Account[]).filter(a => a.type !== 'INCOME' && a.type !== 'EXPENSE'),
        Interval.fromDateTimes(
          DateTime.fromISO('1970'),
          selectedDate,
        ),
      ),
      queryKey,
    ),
    enabled: !!accounts,
    select: select || aggregate,
    networkMode: 'always',
  });

  return result;
}
