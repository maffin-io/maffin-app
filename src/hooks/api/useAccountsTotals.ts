import { DateTime, Interval } from 'luxon';

import type { AccountsTotals } from '@/types/book';
import { useInterval } from '../state';
import { useBalanceSheet } from './useBalanceSheet';
import { useIncomeStatement } from './useIncomeStatement';

/**
 * This is a utility function that returns the balance sheet
 * and the income statement all together. Useful when you want to retrieve
 * the totals for all the accounts (which some are assets/liabilities and
 * others are income/expenses).
 */
export function useAccountsTotals(
  selectedInterval?: Interval,
  select?: (data: AccountsTotals) => AccountsTotals,
): { data: AccountsTotals } {
  const { data: defaultInterval } = useInterval();
  const interval = selectedInterval || defaultInterval;

  const { data: balanceSheet } = useBalanceSheet(
    interval.end as DateTime<true>,
    select,
  );
  const { data: incomeStatement } = useIncomeStatement(interval, select);

  return {
    data: { ...balanceSheet, ...incomeStatement },
  };
}
