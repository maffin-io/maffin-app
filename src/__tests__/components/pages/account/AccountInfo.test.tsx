import React from 'react';
import { render } from '@testing-library/react';
import { DateTime } from 'luxon';
import type { SWRResponse } from 'swr';

import {
  AccountInfo,
  SplitsHistogram,
  TotalLineChart,
} from '@/components/pages/account';
import StatisticsWidget from '@/components/StatisticsWidget';
import * as apiHook from '@/hooks/api';
import type { Account } from '@/book/entities';

jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

jest.mock('@/components/pages/account/SplitsHistogram', () => jest.fn(
  () => <div data-testid="SplitsHistogram" />,
));

jest.mock('@/components/pages/account/TotalLineChart', () => jest.fn(
  () => <div data-testid="TotalLineChart" />,
));

jest.mock('@/components/StatisticsWidget', () => jest.fn(
  () => <div data-testid="StatisticsWidget" />,
));

describe('AccountInfo', () => {
  let account: Account;
  beforeEach(() => {
    jest.spyOn(apiHook, 'useSplits').mockReturnValue({ data: undefined } as SWRResponse);
    jest.spyOn(DateTime, 'now').mockReturnValue(DateTime.fromISO('2023-03-01') as DateTime<true>);
    account = {
      guid: 'guid',
      type: 'ASSET',
      commodity: {
        mnemonic: 'EUR',
      },
    } as Account;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected when no splits', () => {
    const { container } = render(
      <AccountInfo account={account} />,
    );

    expect(TotalLineChart).toBeCalledWith(
      {
        account,
      },
      {},
    );
    expect(SplitsHistogram).toBeCalledWith(
      {
        account,
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'mr-2',
        title: 'You have a total of',
        stats: '€0.00',
        description: 'with an average of €0.00 per month',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      2,
      {
        title: 'This year you have',
        stats: '€0.00',
        description: 'in this account',
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('renders as expected with splits', () => {
    jest.spyOn(apiHook, 'useSplits').mockReturnValue({
      data: [
        {
          transaction: {
            date: DateTime.fromISO('2023-01-01'),
          },
          quantity: 10,
        },
        {
          transaction: {
            date: DateTime.fromISO('2023-03-01'),
          },
          quantity: 50,
        },
      ],
    } as SWRResponse);

    render(
      <AccountInfo account={account} />,
    );

    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'mr-2',
        title: 'You have a total of',
        stats: '€60.00',
        description: 'with an average of -€30.00 per month',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      2,
      {
        title: 'This year you have',
        stats: '€60.00',
        description: 'in this account',
      },
      {},
    );
  });
});
