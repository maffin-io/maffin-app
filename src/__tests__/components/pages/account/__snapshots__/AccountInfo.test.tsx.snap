// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountInfo renders as expected when no splits 1`] = `
<div>
  <div
    class="grid grid-cols-12"
  >
    <div
      class="col-span-6"
    >
      <div
        class="grid grid-cols-12"
      >
        <div
          class="col-span-6"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
        <div
          class="col-span-6"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
        <div
          class="card col-span-12"
        >
          <div
            data-testid="TotalLineChart"
          />
        </div>
      </div>
    </div>
    <div
      class="card col-span-6"
    >
      <div
        class="flex h-full items-center"
      >
        <div
          data-testid="SplitsHistogram"
        />
      </div>
    </div>
  </div>
</div>
`;
